tools/webcam/webcam_apis/nodes/frame_effect_nodes.py

```diff
@@ -0,0 +1,1386 @@
+                 dis_thr: float = 50.0,
+                 flying_heart_dis_thr: float = 100.0):
+
+        super().__init__(
+            name, frame_buffer, output_buffer, enable_key=enable_key)
+
+        if src_img_path is None:
+            src_img_path = 'https://user-images.githubusercontent.com/'\
+                           '87690686/149731850-ea946766-a4e8-4efa-82f5'\
+                           '-e2f0515db8ae.png'
+        if flying_heart_img_path is None:
+            flying_heart_img_path = 'https://user-images.githubusercontent.'\
+                                    'com/87690686/153554948-937ce496-33dd-4'\
+                                    '9ab-9829-0433fd7c13c4.png'
+
+        self.src_img = load_image_from_disk_or_url(src_img_path)
+        self.flying_heart = load_image_from_disk_or_url(flying_heart_img_path)
+
+        self.kpt_thr = kpt_thr
+        self.angle_thr = angle_thr
+        self.dis_thr = dis_thr
+        self.flying_heart_dis_thr = flying_heart_dis_thr
+        self.longest_duration = longest_duration
+        self.largest_size = largest_size
+
+        # record the heart_effect start time for each person
+        self.heart_infos = {}
+        self._heart_start_time = {}
+        self._heart_cur_pos = {}
+        self._heart_tar_pos = {}
+
+    def _cal_distance(self, p1: np.ndarray, p2: np.ndarray) -> np.float64:
+        return np.sqrt((p1[0] - p2[0])**2 + (p1[1] - p2[1])**2)
+
+    # def _cal_angle(self, p1, p2, p3, p4):
+
+    def _cal_angle(self, kpts: np.ndarray, hand_indices: List[int],
+                   finger_indices: List[int]) -> np.float64:
+
+        p1 = kpts[hand_indices[finger_indices[3]]][:2]
+        p2 = kpts[hand_indices[finger_indices[2]]][:2]
+
+        p3 = kpts[hand_indices[finger_indices[1]]][:2]
+        p4 = kpts[hand_indices[finger_indices[0]]][:2]
+
+        v1 = p2 - p1
+        v2 = p4 - p3
+
+        vector_prod = v1[0] * v2[0] + v1[1] * v2[1]
+        length_prod = np.sqrt(pow(v1[0], 2) + pow(v1[1], 2)) * np.sqrt(
+            pow(v2[0], 2) + pow(v2[1], 2))
+        cos = vector_prod * 1.0 / (length_prod * 1.0 + 1e-6)
+
+        return (np.arccos(cos) / np.pi) * 180
+
+    def _check_heart(self, pred: Dict[str, np.ndarray],
+                     hand_indices: List[int], mouth_indices: List[int]) -> int:
+        """Check if a person is posing a 'hand heart' gesture.
+        Args:
+            pred(dict): The pose estimation results containing:
+                - "keypoints" (np.ndarray[K,3]): keypoint detection result
+                                                 in [x, y, score]
+            hand_indices(list[int]): hand keypoint indices.
+
+        Returns:
+            int: a number representing the type of gesture,
+                 0: None, 1: hand heart, 2: left hand blow kiss,
+                 3: right hand blow kiss
+        """
+        kpts = pred['keypoints']
+
+        # note: these indices are corresoponding to the following keypoints:
```

Comment:
Better not hard code the hand keypoint indices here. Consider using a get_xxx_index function, which can be defined at `tools/webcam/webcam_api/utils/pose.py`.